### ISSUE TITLE
Forms created on non-primary monitors with different DPI have incorrect font size

### ISSUE THREAD

**Issue by filipnavara** [dotnet/winforms#6157]
* .NET Core Version: 6.0 RTM

* Have you experienced this same bug with .NET Framework?: No

**Problem description:**

Forms that are launches on secondary screen with different DPI (eg. primary = 150% or 144, secondary = 100% or 96) have incorrectly scaled fonts.

**Expected behavior:**

Correctly scaled fonts.

**Minimal repro:**

[WinFormsApp2.zip](https://github.com/dotnet/winforms/files/7514304/WinFormsApp2.zip)

Run the app, it will likely start on primary monitor. Clicking on the form with create a new form above the parent one. On primary screen it looks like this:

![image](https://user-images.githubusercontent.com/1764393/141152171-6c9fe252-4508-4128-a05d-fcfdfc01b609.png)

Move one of the windows on secondary screen a click on it. The new windows on the secondary screen look like this (upper left one on the screenshot is original form dragged from primary screen):

![image](https://user-images.githubusercontent.com/1764393/141152295-ee0ce2f2-bb0f-4284-b046-369537d4630e.png)

**Comment by filipnavara** [comment]
Ref: https://github.com/dotnet/winforms/issues/4854

**Comment by merriemcgaw** [comment]
@filipnavara is this a regression?

**Comment by filipnavara** [comment]
Yes, it is a regression. You can change the TFM in the app to `net5.0-windows` to see how it behaved previously.

**Comment by dreddy-work** [comment]
@filipnavara, Thanks for reporting this. I am looking into it.

**Comment by dreddy-work** [comment]
Cached values for the property `ScaledControlFont `are indirectly fixed in the #5557. I do not expect this issue with latest bits from 7.0/main branch. I will be working on how to get this serviced. @filipnavara , did you try 7.0 bits here?

**Comment by filipnavara** [comment]
I'll retry with 7.0/main bits tomorrow when I am in the office with multi-monitor setup.

**Comment by filipnavara** [comment]
The font sizing is definitely working better on 7.0. I cannot verify it's 100% correct over VNC due to scaling issues with the remote desktop. Curiously, the forms are still created with a different size based on which displayed they were created on:

![image](https://user-images.githubusercontent.com/1764393/141187385-812b199b-23a3-4f69-91b9-2e4359a78ff6.png)

I think we had a workaround for that in our code but nothing I can verify at the moment or find quickly.

Would it make sense for me to try to backport specific PRs to the 6.0 branch and try running our full app with that?

**Comment by dreddy-work** [comment]
> The font sizing is definitely working better on 7.0. I cannot verify it's 100% correct over VNC due to scaling issues with the remote desktop. Curiously, the forms are still created with a different size based on which displayed they were created on:
> 
> ![image](https://user-images.githubusercontent.com/1764393/141187385-812b199b-23a3-4f69-91b9-2e4359a78ff6.png)
> 
> I think we had a workaround for that in our code but nothing I can verify at the moment or find quickly. At first I thought it's just typical `ClientSize`/`Size` issue but one of the windows is narrower and taller at the same time. I don't have explanation for that.
> 
> Would it make sense for me to try to backport specific PRs to the 6.0 branch and try running our full app with that?

Sure. Or if you can retarget to 7.0?

**Comment by filipnavara** [comment]
> Or if you can retarget to 7.0?

Yeah, I can do that. I'd still need to build SWF myself to get version with 92921bb7f17c576db8bc2eced309c00b0947d344 since it will take some time to propagate to feeds and/or installer package. We have some workarounds that would likely interfere and make it hard to see if it really helped or if we're hitting other issue.

I did initial build with 7.0 (w/o 92921bb7f17c576db8bc2eced309c00b0947d344) and it gave me mixed results. This issue with fonts looks likely to be fixed with 7.0/main but there are two other layout issues in the same control. One of them is related to 92921bb7f17c576db8bc2eced309c00b0947d344. The other one is bug in our code triggered by some other layout quirk which we didn't fully track down yet (but ultimately it's a bug in our code).

**Comment by dreddy-work** [comment]
> there are two other layout issues in the same control. One of them is related to [92921bb](https://github.com/dotnet/winforms/commit/92921bb7f17c576db8bc2eced309c00b0947d344).

If you have more details on other layout issue you are seeing, please share or open an issue against 7.0 and I will track it for 7.0 early preview.  I am getting these two issues you reported serviced in 6.0.

**Comment by filipnavara** [comment]
I hope these two issues are the only major ones. We need to remove couple of workarounds that are not necessary anymore (like our own workaround for initial scaling on non-primary display) and do a little bit more testing. Once I'm done with the testing I'll post an update! Thanks!

**Comment by RussKie** [comment]
> Yeah, I can do that. I'd still need to build SWF myself to get version with [92921bb](https://github.com/dotnet/winforms/commit/92921bb7f17c576db8bc2eced309c00b0947d344) since it will take some time to propagate to feeds and/or installer package.

This guide should help: https://github.com/dotnet/winforms/blob/main/docs/debugging.md#1-copy-your-changes-into-the-sdk
I also have a script that can installs/uninstall dev binaries https://gist.github.com/RussKie/bb11e213f04729603e533856865922c1

**Comment by filipnavara** [comment]
@RussKie Thanks for the link! I'm equipped to make full source builds but this will likely same me a bit of time.

**Comment by filipnavara** [comment]
I can confirm that using build from `main` with the two fixes (92921bb7f17c576db8bc2eced309c00b0947d344 and 9b77d919753564d3a47799c59ca426686e7daf70) resolves the issues in our app. There's one remaining problem with scaling of padding in one of our custom controls but it's very likely in our code and workaroundable/fixable.

**Comment by filipnavara** [comment]
We still found one case which was not addressed by #5557. I'll file a separate issue once we have a repro. The gist is that the Form gets created on non-primary monitor. It arrives to `OnHandleCreated` and the scaled font is cached. Then immediately after that _before the window is shown_ it gets moved to a position on a different monitor (in our case it's done by a component that saves/restores form positions so it's a bit tricky to isolate). It receives `WM_DPICHANGED` event, calls `OnDpiChnaged` and `ScaledControlFont` still has the stale value which causes it to be returned by the `Font` property in the form's `OnFontChanged` method.

**Comment by dreddy-work** [comment]
> I can confirm that using build from `main` with the two fixes ([92921bb](https://github.com/dotnet/winforms/commit/92921bb7f17c576db8bc2eced309c00b0947d344) and [9b77d91](https://github.com/dotnet/winforms/commit/9b77d919753564d3a47799c59ca426686e7daf70)) resolves the issues in our app. There's one remaining problem with scaling of padding in one of our custom controls but it's very likely in our code and workaroundable/fixable.
> 
> Update: I backported both fixed on top of release/6.0 branch (https://github.com/emclient/winforms/tree/release/6.0-em) and now it works flawlessly for our app. Even the padding rescaling problem disappeared once we removed our in-app workarounds.

Very happy to hear that. 92921bb7f17c576db8bc2eced309c00b0947d344 is being back ported. 
 We may not be able to backport 9b77d919753564d3a47799c59ca426686e7daf70 as-is because of risk associated there. I will look into getting a tailored fix to get back to 5.0 behavior there if possible.

**Comment by dreddy-work** [comment]
@filipnavara , [9b77d91](https://github.com/dotnet/winforms/commit/9b77d919753564d3a47799c59ca426686e7daf70) is solving initialization on the secondary monitor and helping scale it when moved back and forth between primary and secondary monitor. Just wanted to check back with you if you had a work around here and can be used for your product? a tailored fix to bring back 5.0 behavior for a Form that was incorrectly initialized on non-primary monitor seems not straight forward.

**Comment by filipnavara** [comment]
We had a workaround for that scenario with .NET 5: https://gist.github.com/filipnavara/0cd86bba0aca62e416d46e19361ec6fa#file-gistfile1-txt-L114-L158. I am pasting the whole file as-is but we already modified it significantly for .NET 6 to remove most of the workarounds: https://gist.github.com/filipnavara/f6d6f7af8102107a56709edd1b00e85d

**Comment by dreddy-work** [comment]
@Olina-Zhang , can you confirm if we can close this with latest 7.0 bits?

**Comment by Philip-Wang01** [comment]
@dreddy-work , verified this issue with latest 7.0 build : 7.0.100-preview.6.22314.18, this issue still repro.
300% main screen: 
![image](https://user-images.githubusercontent.com/86937911/173765121-af72ebb7-2430-4ddb-a346-f07b5bd1c098.png)
Move to 100% sceondary screen to create: 
![image](https://user-images.githubusercontent.com/86937911/173765183-215c25f4-cdff-46fe-aec1-742cb3ef9371.png)

**Comment by dreddy-work** [comment]
> @dreddy-work , verified this issue with latest 7.0 build : 7.0.100-preview.6.22314.18, this issue still repro. 300% main screen: ![image](https://user-images.githubusercontent.com/86937911/173765121-af72ebb7-2430-4ddb-a346-f07b5bd1c098.png) Move to 100% sceondary screen to create: ![image](https://user-images.githubusercontent.com/86937911/173765183-215c25f4-cdff-46fe-aec1-742cb3ef9371.png)

@Philip-Wang01 , i did not get full details from your screen shot. they look right to me. What do you think is wrong here?

**Comment by Philip-Wang01** [comment]
@dreddy-work , The labels' font size is about the same as the scaling of the title font when launching app in secondary screen, but it causes the spacing of the labels to become larger. Is this acceptable behavior?
![image](https://user-images.githubusercontent.com/86937911/173975761-0fbc568d-f249-4099-a6a7-09bb29fb175a.png)

**Comment by dreddy-work** [comment]
> @dreddy-work , The labels' font size is about the same as the scaling of the title font when launching app in secondary screen, but it causes the spacing of the labels to become larger. Is this acceptable behavior?

Hmm. Now i got what you were referring. That may be very well how those labels were placed. This issue was mainly about font sizes and that is resolved. If there is a real customer scenario about spacings going wrong, we will look it separately.
